javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -899,8 +943,22 @@ public int length() {
     public <U> Vector<U> map(Function<? super T, ? extends U> mapper) {
         Objects.requireNonNull(mapper, "mapper is null");
 
-        final Iterator<? extends U> results = iterator().map(mapper);
-        return ofAll(results);
+        final AtomicReference<Object[]> newLeading = new AtomicReference<>(Arrays.empty());
+        final Object[] newMiddle = new Object[middle.leafCount()];
+        final AtomicReference<Object[]> newTrailing = new AtomicReference<>(Arrays.empty());
+        unsafeLeafIterator((leafCount, array, arrayStart, arrayLength) -> {
```

Comment:
> I will come back with a suggestion, please hold your horses.

I made some other modifications also, will push so that you can give the suggestions based on the latest codebase :).

> This is top priority because our users have serious problems.

Understood.